tests/test_coreg.py

```diff
@@ -366,3 +368,101 @@ def test_subsample(self):
         matrix_diff = np.abs(icp_full.to_matrix() - icp_sub.to_matrix())
         # Check that the x/y/z differences do not exceed 30cm
         assert np.count_nonzero(matrix_diff > 0.3) == 0
+
+    def test_apply_matrix(self):
+        warnings.simplefilter("error")
+        # This should maybe be its own function, but would just repeat the data loading procedure..
+
+        # Synthesize a shifted and vertically offset DEM
+        pixel_shift = 11
+        bias = 5
+        shifted_dem = self.ref.data.squeeze().copy()
+        shifted_dem[:, pixel_shift:] = shifted_dem[:, :-pixel_shift]
+        shifted_dem[:, :pixel_shift] = np.nan
+        shifted_dem += bias
+
+        matrix = np.diag(np.ones(4, dtype=float))
+        matrix[0, 3] = pixel_shift * self.tba.res[0]
+        matrix[2, 3] = -bias
+
+        transformed_dem = coreg.apply_matrix(shifted_dem.data.squeeze(), self.ref.transform, matrix)
+
+        diff = np.asarray(self.ref.data.squeeze() - transformed_dem)
+
+        # Check that the median is very close to zero
+        assert np.abs(np.nanmedian(diff)) < 0.05
+        # Check that the NMAD is low
+        assert spatial_tools.nmad(diff) < 3
+
+        def rotation_matrix(rotation=30):
+            rotation = np.deg2rad(rotation)
+            matrix = np.array([
+                [1, 0, 0, 0],
+                [0, np.cos(rotation), -np.sin(rotation), 0],
+                [0, np.sin(rotation), np.cos(rotation), 0],
+                [0, 0, 0, 1]
+            ])
+            return matrix
+
+        rotation = 4
+        rotated_dem = coreg.apply_matrix(
+            self.ref.data.squeeze(),
+            self.ref.transform,
+            rotation_matrix(rotation),
+        )
+        # Make sure that the rotated DEM is way off.
+        assert np.abs(np.nanmedian(rotated_dem - self.ref.data.data)) > 400
+
+        # Apply a rotation in the opposite direction
+        unrotated_dem = coreg.apply_matrix(
+            rotated_dem,
+            self.ref.transform,
+            rotation_matrix(-rotation * 0.989)  # This is not exactly -rotation, probably due to displaced pixels.
```

Comment:
Yeah... I don't know where that is from either...  `apply_matrix()` estimates a deramp. I'd love any suggestions to figure out the cause behind this magic number!

Reply:
Isn't this because of your center of rotation? If I understand your code correctly, the centroid is at the bottom left corner, so this will affect your result.
Should the centroid not be at the center of the DEM?

Reply:
Changed so the centroid is the centre coordinate of the raster bounds in 07bf199.